src/data_structures/fmindex.rs

```diff
@@ -7,58 +7,36 @@
 
 use std::iter::DoubleEndedIterator;
 
-use data_structures::bwt::{Occ, Less, less, BWT};
-use data_structures::suffix_array::SuffixArraySlice;
-use alphabets::{Alphabet, dna};
+use data_structures::bwt::{less, BWT, DerefBWT, DerefOcc, DerefLess};
+use data_structures::suffix_array::{SuffixArray, DerefSuffixArray};
+use alphabets::dna;
 use std::mem::swap;
 
 /// A suffix array interval.
-#[derive(Debug, Copy, Clone)]
+#[derive(Clone, Copy, Debug)]
 pub struct Interval {
-    lower: usize,
-    upper: usize,
+    pub lower: usize,
+    pub upper: usize,
 }
 
-
 impl Interval {
-    /// Return the occurrence positions of the pattern as a slice of the suffix array.
-    pub fn occ<'a>(&self, pos: &'a SuffixArraySlice) -> &'a [usize] {
-        &pos[self.lower..self.upper]
-    }
+  pub fn occ<SA: SuffixArray, T: DerefSuffixArray<SA>>(&self, sa: T) -> Vec<usize> {
```

Comment:
Should it still be generic on the type `SA: SuffixArray` or should I just use `&SuffixArray` and pay the cost of dynamic dispatch? This isn't an issue for the other case since it only accepts a `RawSuffixArray`.